server/vulnerabilities/oval/analyzer.go

```diff
@@ -17,73 +17,124 @@ import (
+
+	var offset int
+	for {
+		hIds, err := ds.HostIDsByOSVersion(ctx, ver, offset, hostsBatchSize)
+		offset += hostsBatchSize
 
-		defs, err := loadDef(platform, vulnPath)
 		if err != nil {
 			errR = multierror.Append(errR, err)
 			continue
 		}
 
-		hIds, err := ds.HostIDsByPlatform(ctx, &v.Platform, &v.Name)
-		if err != nil {
-			errR = multierror.Append(errR, err)
-			continue
+		if len(hIds) == 0 {
+			break
 		}
 
+		foundInBatch := make(map[uint][]fleet.SoftwareVulnerability)
 		for _, hId := range hIds {
-			h := fleet.Host{ID: hId}
-			opts := fleet.SoftwareListOptions{
-				IncludeCVEScores: false,
-				VulnerableOnly:   false,
-				WithHostCounts:   false,
-			}
-			err := ds.LoadHostSoftware(ctx, &h, opts)
+			software, err := ds.ListSoftwareForVulnDetection(ctx, hId)
 			if err != nil {
 				errR = multierror.Append(errR, err)
 				continue
 			}
+			foundInBatch[hId] = defs.Eval(software)
+		}
 
-			found := defs.Eval(h.Software)
-			if len(found) == 0 {
-				continue
-			}
-
-			existing, err := ds.ListSoftwareVulnerabilities(ctx, hId)
-			if err != nil {
-				errR = multierror.Append(errR, err)
-			}
+		existingInBatch, err := ds.ListSoftwareVulnerabilities(ctx, hIds)
+		if err != nil {
+			errR = multierror.Append(errR, err)
+			continue
+		}
 
-			toInsert, toDelete := vulnsDelta(found, existing)
-			if len(toDelete) > 0 {
-				if err := ds.DeleteVulnerabilitiesByCPECVE(ctx, toDelete); err != nil {
-					errR = multierror.Append(errR, err)
-				}
+		for _, hId := range hIds {
+			insrt, del := vulnsDelta(foundInBatch[hId], existingInBatch[hId])
+			for _, i := range insrt {
+				key := fmt.Sprintf("%d:%s", i.SoftwareID, i.CVE)
+				toInsertSet[key] = i
 			}
-			if len(toInsert) > 0 {
-				if _, err = ds.InsertVulnerabilities(ctx, toInsert, fleet.OVAL); err != nil {
-					errR = multierror.Append(errR, err)
-				}
-				return toInsert, errR
+			for _, d := range del {
+				key := fmt.Sprintf("%d:%s", d.SoftwareID, d.CVE)
+				toDeleteSet[key] = d
 			}
 		}
 	}
 
-	return nil, errR
+	batchProcess(toDeleteSet, func(v []fleet.SoftwareVulnerability) {
+		if err := ds.DeleteVulnerabilitiesByCPECVE(ctx, v); err != nil {
+			errR = multierror.Append(errR, err)
+		}
+	})
+
+	inserted := make([]fleet.SoftwareVulnerability, 0, len(toInsertSet))
+	batchProcess(toInsertSet, func(v []fleet.SoftwareVulnerability) {
+		if _, err := ds.InsertVulnerabilities(ctx, v, fleet.OVAL); err != nil {
+			errR = multierror.Append(errR, err)
+		} else {
+			inserted = append(inserted, v...)
+		}
+	})
+
+	return inserted, errR
+}
+
+func batchProcess(
+	values map[string]fleet.SoftwareVulnerability,
+	dsFunc func(v []fleet.SoftwareVulnerability),
+) {
+	if len(values) == 0 {
+		return
+	}
+
+	bSize := vulnBatchSize
+	if bSize > len(values) {
+		bSize = len(values)
+	}
+
+	buffer := make([]fleet.SoftwareVulnerability, bSize)
+
+	var offset, i int
+	for _, v := range values {
+		buffer[offset] = v
```

Comment:
Is there a reason not to use `buffer[i]` here? I think that would have made it clearer to me what was going on in the loop.

Reply:
Oh, I think I understand why now. Perhaps it would be helpful to add a comment in the `if` below?